sched/addrenv/addrenv.c

```diff
@@ -0,0 +1,562 @@
+
+int addrenv_select(FAR struct addrenv_s *addrenv)
+{
+  FAR struct tcb_s *tcb = this_task();
+  addrenv_take(addrenv);
+  tcb->addrenv_curr = addrenv;
+  return addrenv_switch(tcb);
+}
+
+/****************************************************************************
+ * Name: addrenv_restore
+ *
+ * Description:
+ *   Switch back to the procces's own address environment.
+ *
+ * Input Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int addrenv_restore(void)
+{
+  FAR struct tcb_s *tcb = this_task();
+  addrenv_give(tcb->addrenv_curr);
+
+  if (tcb->addrenv_own == NULL)
+    {
+      /* Kernel thread, clear g_addrenv, as it is not valid any more */
+
+      addrenv_clear_current(tcb->addrenv_curr);
+    }
+
+  tcb->addrenv_curr = tcb->addrenv_own;
+  return addrenv_switch(tcb);
+}
+
+/****************************************************************************
+ * Name: addrenv_take
+ *
+ * Description:
+ *   Take a reference to an address environment.
+ *
+ * Input Parameters:
+ *   addrenv - The address environment.
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+void addrenv_take(FAR struct addrenv_s *addrenv)
+{
+  irqstate_t flags = enter_critical_section();
+  addrenv->refs++;
+  leave_critical_section(flags);
+}
+
+/****************************************************************************
+ * Name: addrenv_give
+ *
+ * Description:
+ *   Give back a reference to an address environment.
+ *
+ * Input Parameters:
+ *   addrenv - The address environment.
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
+ *
+ ****************************************************************************/
+
+int addrenv_give(FAR struct addrenv_s *addrenv)
+{
+  irqstate_t flags;
+  int refs;
+
+  flags = enter_critical_section();
+  refs = --addrenv->refs;
+  leave_critical_section(flags);
+
+  return refs;
+}
+
+/****************************************************************************
+ * Name: addrenv_drop
+ *
+ * Description:
+ *   Drop an address environment.
+ *
+ * Input Parameters:
+ *   addrenv - The address environment.
+ *   deferred - yes: The address environment should be dropped by the worker
+ *              no:  The address environment can be dropped at once
+ *
+ * Returned Value:
+ *   This is a NuttX internal function so it follows the convention that
+ *   0 (OK) is returned on success and a negated errno is returned on
+ *   failure.
```

Comment:
Just noticed that the function description is wrong for many of these. Copy&pasted "Returned value" field. I will patch it tomorrow.

Reply:
Done now